src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -623,18 +682,34 @@ private int WriteCommonFields(Span<byte> buffer, TarEntryType actualEntryType)
             {
                 checksum += FormatNumeric(_uid, buffer.Slice(FieldLocations.Uid, FieldLengths.Uid));
             }
+            else if (_uid == 0 && _format is TarEntryFormat.Gnu)
+            {
+                FormatAsZeroChars(buffer.Slice(FieldLocations.Uid, FieldLengths.Uid));
+            }
 
             if (_gid > 0)
             {
                 checksum += FormatNumeric(_gid, buffer.Slice(FieldLocations.Gid, FieldLengths.Gid));
             }
+            else if (_uid == 0 && _format is TarEntryFormat.Gnu)
+            {
+                FormatAsZeroChars(buffer.Slice(FieldLocations.Gid, FieldLengths.Gid));
+            }
 
             if (_size >= 0)
             {
                 checksum += FormatNumeric(_size, buffer.Slice(FieldLocations.Size, FieldLengths.Size));
             }
 
-            checksum += WriteAsTimestamp(_mTime, buffer.Slice(FieldLocations.MTime, FieldLengths.MTime));
+            if (_format is TarEntryFormat.Gnu)
+            {
+                // Just like atime and ctime, mtime should be stored as zeros when the value is UnixEpoch.
+                checksum += WriteAsGnuTimestamp(_mTime, buffer.Slice(FieldLocations.MTime, FieldLengths.MTime));
```

Comment:
I don't understand why we need a specific method `WriteAsGnuTimestamp`. It seems to handle the `DateTimeOffset.UnixEpoch` separate from any other timestamp?

I don't think such handling is necessary and that we should be able to remove that method and call `WriteAsTimestamp` directly.

Reply:
I know `WriteAsTimestamp` has an `else` branch to check for Gnu, but when debugging I noticed we were not entering it for this case. I didn't want to mess with that code at the moment because even the slightest change was making several unrelated unit tests fail, so I decided to create a separate method for the Gnu timestamps.

See my main answer, I'd welcome addressing this better in a separate PR.

Reply:
Method removed. We can keep using WriteAsTimestamp.